pgmax/fg/graph.py

```diff
@@ -160,51 +157,79 @@ def _register_factor_group(
                 f"A factor group with the name {name} already exists. Please choose a different name!"
             )
 
-        self._factor_group_to_msgs_starts[factor_group] = self._total_factor_num_states
-        self._factor_group_to_potentials_starts[
-            factor_group
-        ] = self._total_factor_num_configs
-        factor_num_states_cumsum = np.insert(
-            factor_group.factor_num_states.cumsum(), 0, 0
-        )
-        factor_group_num_configs = 0
-        for vv, variables in enumerate(factor_group._variables_to_factors):
+        for variables in factor_group._variables_to_factors:
             if variables in self._variables_to_factors:
                 raise ValueError(
                     f"A factor involving variables {variables} already exists. Please merge the corresponding factors."
                 )
 
             factor = factor_group._variables_to_factors[variables]
             self._variables_to_factors[variables] = factor
-            self._factor_to_msgs_starts[factor] = (
-                self._factor_group_to_msgs_starts[factor_group]
-                + factor_num_states_cumsum[vv]
-            )
-            self._factor_to_potentials_starts[factor] = (
-                self._factor_group_to_potentials_starts[factor_group]
-                + vv * factor.log_potentials.shape[0]
-            )
-            factor_group_num_configs += factor.log_potentials.shape[0]
 
-        self._total_factor_num_states += factor_num_states_cumsum[-1]
-        self._total_factor_num_configs += factor_group_num_configs
+            factor_subtype = factor.__subtype__
+            if factor_subtype not in self._subtypes_to_factors:
```

Comment:
Instead of this if, we should have a centralized list of all registered factor types right now, and when initializing the factor graph, in `__post_init__` we should already initialize `self._factor_type_to_factors = {factor_type: collections.OrderedDict()}`.

The idea is that all returned dictionaries would have one entry per factor type, regardless of whether there is a factor of that type in the factor graph. It just makes things uniform and reduces special processing in the form of this kind of ifs.